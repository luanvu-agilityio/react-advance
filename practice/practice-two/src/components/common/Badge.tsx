import type { ReactNode } from 'react'
import styled from 'styled-components'

interface BadgeProps {
  children: ReactNode
}

const Badge = styled.span`
  display: inline-block;
  padding: 2.5px 7.5px;
  font-size: 12px;
  font-weight: var(--font-weight-semibold);
  border-radius: 12px;
  background-color: var(--green-shade-4);
  color: var(--green-color-default);
`
export const BadgeComponent = ({ children }: BadgeProps) => {
  return <Badge>{children}</Badge>
}
