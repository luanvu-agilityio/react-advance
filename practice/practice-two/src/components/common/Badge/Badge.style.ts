import styled from 'styled-components'

export const BadgeContainer = styled.span`
  display: inline-block;
  padding: 2.5px 7.5px;
  font-size: 12px;
  font-weight: var(--font-weight-semibold);
  border-radius: 12px;
  background-color: var(--green-shade-4);
  color: var(--green-color-default);
`
