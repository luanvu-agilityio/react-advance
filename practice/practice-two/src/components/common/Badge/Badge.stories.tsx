import type { Meta, StoryObj } from '@storybook/react'
import Badge from './index'

const meta: Meta<typeof Badge> = {
  title: 'Components/Common/Badge',
  component: Badge,
  parameters: {
    layout: 'centered',
    docs: {
      description: {
        component:
          'A versatile badge component for displaying status, labels, or categories with consistent styling.',
      },
    },
  },
  argTypes: {
    children: {
      description: 'Content to display inside the badge',
      control: 'text',
    },
  },
  tags: ['autodocs'],
}

export default meta
type Story = StoryObj<typeof Badge>

export const Default: Story = {
  args: {
    children: 'New',
  },
}

export const WithText: Story = {
  args: {
    children: 'Featured',
  },
  parameters: {
    docs: {
      description: {
        story: 'Badge displaying simple text content',
      },
    },
  },
}

export const WithNumber: Story = {
  args: {
    children: '5',
  },
  parameters: {
    docs: {
      description: {
        story: 'Badge displaying numeric content like counts or quantities',
      },
    },
  },
}

export const WithPercentage: Story = {
  args: {
    children: '25% OFF',
  },
  parameters: {
    docs: {
      description: {
        story: 'Badge displaying discount or percentage information',
      },
    },
  },
}

export const WithEmoji: Story = {
  args: {
    children: 'ðŸ”¥ Hot',
  },
  parameters: {
    docs: {
      description: {
        story: 'Badge with emoji and text content',
      },
    },
  },
}

export const LongText: Story = {
  args: {
    children: 'Limited Time Offer',
  },
  parameters: {
    docs: {
      description: {
        story:
          'Badge with longer text content to test text wrapping and sizing',
      },
    },
  },
}

export const SingleCharacter: Story = {
  args: {
    children: 'A',
  },
  parameters: {
    docs: {
      description: {
        story: 'Badge with single character content',
      },
    },
  },
}

export const WithSpecialCharacters: Story = {
  args: {
    children: 'â˜…â˜…â˜…â˜…â˜…',
  },
  parameters: {
    docs: {
      description: {
        story: 'Badge displaying special characters like stars or symbols',
      },
    },
  },
}

export const StatusBadges: Story = {
  render: () => (
    <div style={{ display: 'flex', gap: '8px', flexWrap: 'wrap' }}>
      <Badge>New</Badge>
      <Badge>Sale</Badge>
      <Badge>Popular</Badge>
      <Badge>Limited</Badge>
      <Badge>Bestseller</Badge>
    </div>
  ),
  parameters: {
    docs: {
      description: {
        story: 'Multiple badges showing different status indicators',
      },
    },
  },
}

export const CountBadges: Story = {
  render: () => (
    <div style={{ display: 'flex', gap: '8px', alignItems: 'center' }}>
      <span>Notifications</span>
      <Badge>3</Badge>
      <span>Cart</span>
      <Badge>12</Badge>
      <span>Messages</span>
      <Badge>99+</Badge>
    </div>
  ),
  parameters: {
    docs: {
      description: {
        story: 'Badges used as notification counters alongside labels',
      },
    },
  },
}

export const EmptyBadge: Story = {
  args: {
    children: '',
  },
  parameters: {
    docs: {
      description: {
        story: 'Edge case: badge with empty content',
      },
    },
  },
}
