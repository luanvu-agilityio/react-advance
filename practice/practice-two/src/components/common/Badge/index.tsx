import type { ReactNode } from 'react'
import { BadgeContainer } from './Badge.style'

export interface BadgeProps {
  children: ReactNode
}

const Badge = ({ children }: BadgeProps) => {
  return <BadgeContainer>{children}</BadgeContainer>
}
export default Badge
