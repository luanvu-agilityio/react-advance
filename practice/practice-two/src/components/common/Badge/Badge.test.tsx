import { render } from '@testing-library/react'
import Badge from './index'

describe('Badge', () => {
  // Basic rendering snapshot
  it('renders basic content correctly', () => {
    const { container: stringBadge } = render(<Badge>Test Badge</Badge>)
    expect(stringBadge).toMatchSnapshot('string-content')

    const { container: numericBadge } = render(<Badge>42</Badge>)
    expect(numericBadge).toMatchSnapshot('numeric-content')

    const { container: zeroBadge } = render(<Badge>{0}</Badge>)
    expect(zeroBadge).toMatchSnapshot('zero-content')
  })

  // Different content types snapshot
  it('renders different content types correctly', () => {
    const { container: specialCharBadge } = render(<Badge>★★★★★</Badge>)
    expect(specialCharBadge).toMatchSnapshot('special-characters')

    const { container: htmlBadge } = render(
      <Badge>
        <strong>Bold</strong> Text
      </Badge>
    )
    expect(htmlBadge).toMatchSnapshot('html-content')
  })

  // Edge cases snapshot
  it('handles long text correctly', () => {
    const longText =
      'This is a very long badge text that should still render properly'
    const { container: longTextBadge } = render(<Badge>{longText}</Badge>)
    expect(longTextBadge).toMatchSnapshot('long-text')

    const { container: multipleBadges } = render(
      <div>
        <Badge>Badge 1</Badge>
        <Badge>Badge 2</Badge>
        <Badge>Badge 3</Badge>
      </div>
    )
    expect(multipleBadges).toMatchSnapshot('multiple-badges')
  })
})
