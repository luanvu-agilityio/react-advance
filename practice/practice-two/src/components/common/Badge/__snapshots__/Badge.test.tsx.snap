// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge handles edge cases correctly: long-text 1`] = `
<div>
  <span
    class="sc-Qotzb iVVyXB"
  >
    This is a very long badge text that should still render properly
  </span>
</div>
`;

exports[`Badge handles edge cases correctly: multiple-badges 1`] = `
<div>
  <div>
    <span
      class="sc-Qotzb iVVyXB"
    >
      Badge 1
    </span>
    <span
      class="sc-Qotzb iVVyXB"
    >
      Badge 2
    </span>
    <span
      class="sc-Qotzb iVVyXB"
    >
      Badge 3
    </span>
  </div>
</div>
`;

exports[`Badge handles long text correctly: long-text 1`] = `
<div>
  <span
    class="sc-Qotzb iVVyXB"
  >
    This is a very long badge text that should still render properly
  </span>
</div>
`;

exports[`Badge handles long text correctly: multiple-badges 1`] = `
<div>
  <div>
    <span
      class="sc-Qotzb iVVyXB"
    >
      Badge 1
    </span>
    <span
      class="sc-Qotzb iVVyXB"
    >
      Badge 2
    </span>
    <span
      class="sc-Qotzb iVVyXB"
    >
      Badge 3
    </span>
  </div>
</div>
`;

exports[`Badge renders basic content correctly: numeric-content 1`] = `
<div>
  <span
    class="sc-Qotzb iVVyXB"
  >
    42
  </span>
</div>
`;

exports[`Badge renders basic content correctly: string-content 1`] = `
<div>
  <span
    class="sc-Qotzb iVVyXB"
  >
    Test Badge
  </span>
</div>
`;

exports[`Badge renders basic content correctly: zero-content 1`] = `
<div>
  <span
    class="sc-Qotzb iVVyXB"
  >
    0
  </span>
</div>
`;

exports[`Badge renders different content types correctly: html-content 1`] = `
<div>
  <span
    class="sc-Qotzb iVVyXB"
  >
    <strong>
      Bold
    </strong>
     Text
  </span>
</div>
`;

exports[`Badge renders different content types correctly: special-characters 1`] = `
<div>
  <span
    class="sc-Qotzb iVVyXB"
  >
    ★★★★★
  </span>
</div>
`;
